cargo-guppy/src/lib.rs

```diff
@@ -18,18 +25,136 @@ pub fn cmd_diff(json: bool, old: &str, new: &str) -> Result<(), Error> {
+    include_build: bool,
+    omit_edges_into: &[String],
+    roots: &[String],
+) -> Result<(), anyhow::Error> {
+    let mut pkg_graph = PackageGraph::default();
+    let mut package_ids = HashSet::new();
+    let mut omitted_package_ids = HashSet::new();
+
+    let root_set: HashSet<String> = if roots.len() > 0 {
+        roots.iter().cloned().collect()
+    } else {
+        pkg_graph
+            .packages()
+            .filter_map(|meta| {
+                if meta.in_workspace() {
+                    Some(meta.name().to_string())
+                } else {
+                    None
+                }
+            })
+            .collect()
+    };
+
+    let omitted_set: HashSet<String> = omit_edges_into.iter().cloned().collect();
+
+    for metadata in pkg_graph.packages() {
+        if root_set.contains(metadata.name()) {
+            package_ids.insert(metadata.id().clone());
+        }
+        if omitted_set.contains(metadata.name()) {
+            omitted_package_ids.insert(metadata.id().clone());
+        }
+    }
+
+    let transitive_deps: Vec<_> = pkg_graph
+        .select_forward(&package_ids)?
+        .into_iter_ids(None)
+        .chain(&package_ids)
+        .cloned()
+        .collect();
+
+    pkg_graph.retain_edges(|_, DependencyLink { from, to, edge }| {
+        let include_kind = match kind {
+            Kind::All | Kind::ThirdParty => true,
+            Kind::Workspace => from.in_workspace() && to.in_workspace(),
+        };
+
+        let include_target = if let Some(ref target) = target {
+            edge.normal()
+                .and_then(|meta| meta.target())
+                .and_then(|edge_target| {
+                    let res = target_spec::eval(edge_target, target).unwrap_or(true);
+                    Some(res)
+                })
+                .unwrap_or(true)
+        } else {
+            true
+        };
+
+        let include_type = edge.normal().is_some()
+            || include_dev && edge.dev().is_some()
+            || include_build && edge.build().is_some();
+
+        let include_dep = transitive_deps.contains(from.id()) && transitive_deps.contains(to.id());
+
+        let include_edge = !omitted_package_ids.contains(to.id());
+
+        include_kind && include_target && include_type && include_edge && include_dep
+    });
+
+    for package in pkg_graph.select_forward(&package_ids)?.into_iter_ids(None) {
+        let in_workspace = pkg_graph.metadata(package).unwrap().in_workspace();
+        let show_package = match kind {
+            Kind::All => true,
+            Kind::Workspace => in_workspace,
+            Kind::ThirdParty => !in_workspace,
+        };
+        if show_package {
+            println!("{}", pkg_graph.metadata(package).unwrap().id());
+        }
+    }
+
+    if let Some(ref output_file) = output {
+        let dot = pkg_graph
+            .select_forward(&package_ids)?
+            .into_dot(NameVisitor);
+        fs::write(output_file, format!("{}", dot))?;
```

Comment:
you can probably do this with less in-memory buffering:

```
let f = File::create(output_file);
write!(f, "{}", dot)?;
```

Reply:
Changed.